staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -0,0 +1,354 @@
+
+	server := newApfServerWithSingleRequest(decisionSkipFilter, t)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces?watch=true", server.URL), http.StatusOK); err != nil {
+		// request should not be rejected
+		t.Error(err)
+	}
+}
+
+func TestApfRejectRequest(t *testing.T) {
+	epmetrics.Register()
+
+	server := newApfServerWithSingleRequest(decisionReject, t)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusTooManyRequests); err != nil {
+		t.Error(err)
+	}
+
+	checkForExpectedMetricsWithRetry(t, []string{
+		"apiserver_request_terminations_total",
+		"apiserver_dropped_requests_total",
+	})
+}
+
+func TestApfExemptRequest(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+
+	// Wait for at least one sampling window to pass since creation of metrics.ReadWriteConcurrencyObserverPairGenerator,
+	// so that an observation will cause some data to go into the Prometheus metrics.
+	time.Sleep(time.Millisecond * 50)
+
+	server := newApfServerWithSingleRequest(decisionNoQueuingExecute, t)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusOK); err != nil {
+		t.Error(err)
+	}
+
+	checkForExpectedMetricsWithRetry(t, []string{
+		"apiserver_current_inflight_requests",
+		"apiserver_flowcontrol_read_vs_write_request_count_watermarks",
+		"apiserver_flowcontrol_read_vs_write_request_count_samples",
+	})
+}
+
+func TestApfExecuteRequest(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+
+	// Wait for at least one sampling window to pass since creation of metrics.ReadWriteConcurrencyObserverPairGenerator,
+	// so that an observation will cause some data to go into the Prometheus metrics.
+	time.Sleep(time.Millisecond * 50)
+
+	server := newApfServerWithSingleRequest(decisionQueuingExecute, t)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusOK); err != nil {
+		t.Error(err)
+	}
+
+	checkForExpectedMetricsWithRetry(t, []string{
+		"apiserver_current_inflight_requests",
+		"apiserver_current_inqueue_requests",
+		"apiserver_flowcontrol_read_vs_write_request_count_watermarks",
+		"apiserver_flowcontrol_read_vs_write_request_count_samples",
+	})
+}
+
+func TestApfExecuteMultipleRequests(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+
+	// Wait for at least one sampling window to pass since creation of metrics.ReadWriteConcurrencyObserverPairGenerator,
+	// so that an observation will cause some data to go into the Prometheus metrics.
+	time.Sleep(time.Millisecond * 50)
+
+	concurrentRequests := 5
+	preWait := make([]sync.WaitGroup, 4)
+	for i := 0; i < 4; i++ {
+		preWait[i] = sync.WaitGroup{}
+		preWait[i].Add(concurrentRequests)
+	}
+
+	postWait := make([]sync.WaitGroup, 3)
+	for i := 0; i < 3; i++ {
+		postWait[i] = sync.WaitGroup{}
+		postWait[i].Add(concurrentRequests)
+	}
```

Comment:
The members of these arrays are all used in individual ways.  We would normally declare individuals in this case.
```suggestion
	var preStartExecute, postStartExecute, preEnqueue, postEnqueue, preDequeue, postDequeue, finishExecute sync.WaitGroup
	for _, wg := range ([]*sync.WaitGroup{&preStartExecute, &postStartExecute, &preEnqueue, &postEnqueue, &preDequeue, &postDequeue, &finishExecute}) {
	  wg.Add(concurrentRequests)
	}
```

Reply:
@MikeSpreitzer Thank you for the suggestions and the detailed explanation!